hazelcast-client/src/main/java/com/hazelcast/client/impl/HazelcastClientInstanceImpl.java

```diff
@@ -289,18 +304,113 @@ private MetricsRegistryImpl initMetricsRegistry() {
         }
 
+        ClientAwsConfig awsConfig = networkConfig.getAwsConfig();
+        AwsAddressProvider awsAddressProvider = initAwsAddressProvider(awsConfig);
+        if (awsAddressProvider != null) {
+            addressProviders.add(awsAddressProvider);
+        }
+
+        ClientCloudConfig cloudConfig = networkConfig.getCloudConfig();
+        HazelcastCloudAddressProvider cloudAddressProvider = initCloudAddressProvider(cloudConfig);
+        if (cloudAddressProvider != null) {
+            addressProviders.add(cloudAddressProvider);
+        }
+
+        addressProviders.add(new DefaultAddressProvider(networkConfig, addressProviders.isEmpty()));
+        return addressProviders;
+    }
+
+    private HazelcastCloudAddressProvider initCloudAddressProvider(ClientCloudConfig cloudConfig) {
+        if (cloudConfig.isEnabled()) {
+            String discoveryToken = cloudConfig.getDiscoveryToken();
+            return new HazelcastCloudAddressProvider(discoveryToken, getConnectionTimeoutMillis(), loggingService);
+        }
+
+        String cloudToken = properties.getString(ClientProperty.HAZELCAST_CLOUD_DISCOVERY_TOKEN);
+        if (cloudToken != null) {
+            return new HazelcastCloudAddressProvider(cloudToken, getConnectionTimeoutMillis(), loggingService);
+        }
+        return null;
+    }
+
+    private AwsAddressProvider initAwsAddressProvider(ClientAwsConfig awsConfig) {
         if (awsConfig != null && awsConfig.isEnabled()) {
             try {
-                addressProviders.add(new AwsAddressProvider(awsConfig, loggingService));
+                return new AwsAddressProvider(awsConfig, loggingService);
             } catch (NoClassDefFoundError e) {
                 ILogger logger = loggingService.getLogger(HazelcastClient.class);
                 logger.warning("hazelcast-aws.jar might be missing!");
                 throw e;
             }
         }
+        return null;
+    }
 
-        addressProviders.add(new DefaultAddressProvider(networkConfig, addressProviders.isEmpty()));
-        return addressProviders;
+    private AddressTranslator createAddressTranslator() {
+        ClientNetworkConfig networkConfig = getClientConfig().getNetworkConfig();
+        ClientAwsConfig awsConfig = networkConfig.getAwsConfig();
+        ClientCloudConfig cloudConfig = networkConfig.getCloudConfig();
+
+        List<String> addresses = networkConfig.getAddresses();
+        boolean addressListProvided = addresses.size() != 0;
+        boolean awsDiscoveryEnabled = awsConfig != null && awsConfig.isEnabled();
+        boolean discoverySpiEnabled = discoveryService != null;
+        String cloudDiscoveryToken = properties.getString(HAZELCAST_CLOUD_DISCOVERY_TOKEN);
+        if (cloudDiscoveryToken != null && cloudConfig.isEnabled()) {
+            throw new IllegalStateException("Ambiguous hazelcast.cloud configuration. "
+                    + "Both property and configuration provided together. Use only one.");
```

Comment:
better to rephrase as: "Both property based and client configuration based settings are provided for Hazelcast cloud discovery together. Use only one."